pkg/skaffold/log/provider.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package log
+
+import (
+	"sync"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/logger"
+)
+
+type Provider interface {
+	Get() Logger
+}
+
+type fullProvider struct {
+	tail bool
+
+	kubernetesLogger *logger.LogAggregator
+	noopLogger       *NoopLogger
+}
+
+var (
+	provider *fullProvider
+	once     sync.Once
+)
+
+func NewLogProvider(config logger.Config, cli *kubectl.CLI, podSelector kubernetes.PodSelector) Provider {
+	once.Do(func() {
+		kLog := logger.NewLogAggregator(cli, podSelector, config)
+		provider = &fullProvider{
+			tail:             config.Tail(),
+			kubernetesLogger: kLog,
+			noopLogger:       &NoopLogger{},
+		}
```

Comment:
IIUC, there is one instance of k8 logger shared amongst all 3 types of deployer right now. 
1) kubectl, 
2) kustomize and
3) helm.
I wonder if it makes sense to have 3 separate k8 logger instances with correct list of podSelector so only pods deployed by corresponding logger. 

This would make things easier and no syncs needed. You can call stop and start for individual deployers.

Reply:
The current pod selector just searches for the artifact image in the set of containers in a pod. So it's possible that if the same artifact is used in multiple deployments, then the logs will be duplicated if the k8s logger is not a singleton.

Reply:
@gsquared94 came to this exact revelation when playing with this again yesterday. if we move the `podSelector` inside the `Deployer`, we can have separate `Logger` implementations per `Deployer` and the `Runner` doesn't need to know anything about them. will send a separate PR for this and then update this one.